File: betabot_obstacles_avoider/src/move.py
#!/usr/bin/env python

import rospy
import math
import random as rand

from geometry_msgs.msg import Twist
from sensor_msgs.msg import LaserScan
from std_msgs.msg import Float32

range_front = Float32()
range_left = Float32()
range_right = Float32()
#Initial values

normal_vel =0.5
max_angle = 1.57284
min_angle = -max_angle

def callback(msg):
	# rospy.loginfo("MAX LEN %f",len(msg.ranges))
	range_right.data= min(min(msg.ranges[0:999]), 10)
	# range_front.data= min(min(msg.ranges[666:1332]), 10)
	range_left.data= min(min(msg.ranges[1000:1999]), 10)


def move():
	# Initialize node
	rospy.init_node('laser')
	pub = rospy.Publisher('cmd_vel', Twist, queue_size=5)
	sub = rospy.Subscriber("scan", LaserScan, callback)
	rate = rospy.Rate(10) # 10hz
	twist = Twist()
	# Move the robot
	while not rospy.is_shutdown():
		rospy.loginfo("VALUE LEFT %f",range_left.data)
		# rospy.loginfo("VALUE FRONT %f", range_front.data)
		rospy.loginfo("VALUE RIGHT %f", range_right.data)
		if (range_right.data >0.5):
			twist.linear.x = normal_vel; twist.linear.y = normal_vel; twist.linear.z = 0
			twist.angular.x = 0; twist.angular.y = 0; twist.angular.z = 0
			pub.publish(twist)
			rospy.loginfo("Moving right with speed %f", normal_vel)

		else:
			if (range_left.data>0.5):
				twist.linear.x = normal_vel; twist.linear.y = -normal_vel; twist.linear.z = 0
				twist.angular.x = 0; twist.angular.y = 0; twist.angular.z =  normal_vel
				pub.publish(twist)
				rospy.loginfo("Obstacle on my right! moving left %f", -normal_vel)
			else:
				twist.linear.x =0; twist.linear.y = 0; twist.linear.z = 0
				twist.angular.x = 0; twist.angular.y = 0; twist.angular.z =  normal_vel
				pub.publish(twist)
				rospy.loginfo("Obstacle at my left! rotating right %f", twist.angular.z)
		rate.sleep()
		

if __name__ == '__main__':
    try:
        move()
    except rospy.ROSInterruptException:
        pass

File: betabot_obstacles_avoider/src/move.cpp

#include "ros/ros.h"
#include "sensor_msgs/LaserScan.h"
#include <geometry_msgs/Twist.h>
#include <stdio.h>
#define _USE_MATH_DEFINES
#include <cmath>

class ObstacleAvoider{

public:
	ros::Publisher vel_pub;

	void Callback(const sensor_msgs::LaserScan::ConstPtr& msg)
	{
		geometry_msgs::Twist cmdvel;
	// printf("Position: [%f] [%f]\n", msg->range_min,msg->range_max);
		min_dist=msg->ranges[0];
		min_angle=0;
		//Min angle is 0, max is 2 pi
		for(int j=0;j<=1999;j++)
			{
				if(msg->ranges[j]<min_dist)
				{
					min_dist=msg->ranges[j];
					min_angle=j*2*M_PI/2000;
				}
			}
			// printf("max angle is %f  and  min angle is %f", msg->angle_max, msg->angle_min);
			printf("minimum distance is [%f] at an angle of [%f]\n",min_dist,min_angle);
		if(min_dist<=1.0)  // min_dist<=0.5 gave box pushing like behaviour, min_dist<=1.2 gave obstacle avoidance
		{
			if(min_angle<90)
			{
				cmdvel.angular.z=1.0;
				cmdvel.linear.x=0;
				
				cmdvel.linear.y=0;
				cmdvel.linear.z=0;
				cmdvel.angular.x=0;
				cmdvel.angular.y=0;
				printf("left\n");
			}
			else
			{
				cmdvel.angular.z=-1.0;
				cmdvel.linear.x=0;

				cmdvel.linear.y=0;
				cmdvel.linear.z=0;
				cmdvel.angular.x=0;
				cmdvel.angular.y=0;
				printf("right\n");
			}
		}
		else
		{
			cmdvel.linear.x=1.0;
			cmdvel.angular.z=0;
			printf("straight\n");
		}
		
		vel_pub.publish(cmdvel);

	}
	private:
	float min_dist;
	float min_angle;

};

int main(int argc, char **argv)
{
  
  ros::init(argc, argv, "laser");
  ros::NodeHandle nh;

  ObstacleAvoider avoider;

  avoider.vel_pub = nh.advertise<geometry_msgs::Twist>("cmd_vel",1);
  ros::Subscriber sub = nh.subscribe("scan", 1, &ObstacleAvoider::Callback, &avoider);

  ros::spin();

  return 0;
}